drone/src/agent/executor.rs

```diff
@@ -63,9 +68,10 @@ impl Executor {
             database,
             nc,
             _container_events_handle: Arc::new(container_events_handle),
+            backend_to_monitor: Arc::default(),
```

Comment:
So, the idea here is that backend_to_monitor lives as long as someone has a ref to it right? Shouldn't we be able to rework this into a non refcounted type which drops when container reaches a nonliving state?

No change needed for this PR, I'm just wondering.

Reply:
Yes, it lives as long as there is an `Executor`. It's necessary here because we clone the Executor, but we never actually access `backend_to_monitor` on those clones, so we could refactor to break `Executor` up into a smaller struct that gets passed around and remove this `Arc`.

We could design the control loops to exit when they reach a terminal state, in which case we could just spawn them and leave them unattached (we would drop the `JoinHandle`, which does not abort the underlying task), and not have a problem. The advantages to doing it this way are:

- We ensure that only one of each control loop is running per container. All control loops are owned by a `BackendMonitor`, and a `BackendMonitor` is only kept around by putting it into the "slot" belonging to that backend in `backend_to_monitor`. If we create a second `BackendMonitor` for the same backend, as soon as we insert it into `backend_to_monitor`, the old `BackendMonitor` is dropped. Because it is implemented to abort tasks on dropping, those redundant control loops get aborted.
- It makes it harder to create leaks where control loops aren't terminated after a container exits, because instead of verifying that each control loop self-aborts, we just need to verify that `BackendMonitor` gets dropped.

Reply:
I'm not sure what you mean by having a futures queue in the executor? Do you mean flattening it into a state machine to avoid calling `tokio::spawn` entirely?

Reply:
yes but no state machine, just spawn_queue: Queue<tokio::Futures> field in the Executor struct, so start_backend and kill_backend instead of returning futures, return immediately after calling Executor.push_future_onto_queue(docker.kill_backend(...))
kinda like io_uring. But then it'll need callbacks for errors and such so maybe not. There's probably a more ergonomic rust way to achieve that.